classic/selection/FeatureModel.js

```diff
@@ -0,0 +1,238 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+  * A row selection model which enables automatic selection of features
+  * in the map when rows are selected in the grid and vice-versa.
+  *
+  * @class GeoExt.selection.FeatureModel
+  */
+Ext.define('GeoExt.selection.FeatureModel', {
+    extend: 'Ext.selection.RowModel',
+    alias: 'selection.featuremodel',
+
+    /**
+     * The currently selected features.
+     * @property {ol.Collection<ol.Feature>}
+     */
+    selectedFeatures: null,
+
+    /**
+     * The connected vector layer.
+     * @cfg {ol.layer.Vector}
+     * @property {ol.layer.Vector}
+     */
+    layer: null,
+
+    /**
+     * The OpenLayers map we work on
+     * @cfg {ol.Map}
+     */
+    map: null,
+
+    /**
+     * Set to true to create a click handler on the map selecting a clicked
+     * object in the #layer.
+     * @cfg {Boolean}
+     */
+    mapSelection: false,
+
+    /**
+     * The default style for the selected features.
+     * @cfg {ol.style.Style}
+     */
+    selectStyle: new ol.style.Style({
+        image: new ol.style.Circle({
+            radius: 6,
+            fill: new ol.style.Fill({
+                color: 'rgba(255,255,255,0.8)'
+            }),
+            stroke: new ol.style.Stroke({
+                color: 'darkblue',
+                width: 2
+            })
+        }),
+        fill: new ol.style.Fill({
+            color: 'rgba(255,255,255,0.8)'
+        }),
+        stroke: new ol.style.Stroke({
+            color: 'darkblue',
+            width: 2
+        })
+    }),
+
+    /**
+     * The attribute key to mark an OL feature as selected.
+     * @cfg {String}
+     * @property
+     * @readonly
+     */
+    selectedFeatureAttr: 'gx_selected',
+
+    /**
+     * Prepare several connected objects once the selection model is ready.
+     *
+     * @private
+     */
+    bindComponent: function() {
+        var me = this;
+
+        me.callParent(arguments);
+
+        me.selectedFeatures = new ol.Collection();
+
+        // detect a layer from the store if not passed in
+        if (!me.layer || !me.layer instanceof ol.layer.Vector) {
+            var store = me.getStore();
+            if (store && store.getLayer() &&
+                store.getLayer() instanceof ol.layer.Vector) {
+                me.layer = store.getLayer();
+            }
+        }
+
+        // bind several OL events
+        me.bindOlEvents();
+    },
+
+    /**
+     * Binds several events on the OL objects used in this class.
+     *
+     * @private
+     */
+    bindOlEvents: function() {
```

Comment:
We should probably also unregister the listeners on all those objects, when we are destroyed.